wiredtiger_open configuration rework.

keithbostic opened the pull request:
`wiredtiger_open` configuration rework, reference 7abf905.

keithbostic commented:
@michaelcahill, @sueloverso:

Michael, Sue was trying to figure out how to strip `encryption=(secretkey=)` from the base configuration file and she and I started talking about it (there's some discussion in 7abf905).

I started looking at it, and wandering down the rabbit hole, and here's where I ended up.

c56f9eb is a bug fix to the shared-cache configuration processing.

ab5443a is a simplification to the code that creates the base configuration file (and keeps configuration strings from showing up more than once in that file).

071d545 and fb4a089 are changes to make sure there's always a version in the configuration stack (previously, we only had a version if one appeared in the base configuration file).

c108490 and 7aee6d5 are the bigger change:

* Instead of passing around the cfg array for `wiredtiger_open`, I do a merge of `cfg[0-N]` as soon as we have all of the configuration information, that is, I create `WT_CONNECTION.cfg` immediately and use it from that point on.
* Additionally, I do a merge of `cfg[1-N]` into a new string, which is what we'll write into the base configuration file.
* Finally, I changed the config merge code to allow values to be stripped out of config stack. This is how Sue (finally!) gets what she needs, we can strip out any configuration strings we don't want as part of a merge.

I almost wrote a routine to strip configuration values out of strings, but that would have been pretty complex, and it wasn't difficult to add this functionality to the configuration merge functions. I don't think removing configuration values will be heavily used in the future, so I don't see a reason to optimize it at this point.

Michael, if you approve of these changes in principle, Sue, would you please do the review?

agorrod commented on src/conn/conn_cache_pool.c:
```diff
@@ -154,7 +154,7 @@ __wt_cache_pool_config(WT_SESSION_IMPL *session, const char **cfg)
 	 *  creating the pool) - use the chunk size; that's the default.
 	 */
-	if (__wt_config_gets(session, &cfg[1],
+	if (__wt_config_gets(session, cfg,
```
Can you explain this change? The code was trying to pull just the user-specified value and exclude the base configuration.

keithbostic replied:
Maybe I'm missing something, but why would this code exclude the base configuration?

Note 25 lines before, in the `!created` code, it's using `cfg`, not `&cfg[1]` -- why is this code, which is executed in that case, using `&cfg[1]`?

agorrod commented:
Thanks @keithbostic The cache pool stuff looks good to me now.

keithbostic commented:
@michaelcahill, @sueloverso, @agorrod: Alex talked me out of c56f9eb, and I've reverted it -- the shared cache code requires cfg[1] to be the non-standard information, even during `wiredtiger_open`.

To make that continue to work, I changed `wiredtiger_open` (2b07239), so instead of `cfg[0] = conn->cfg`, I do `cfg[0] = base-config, cfg[1] = base_merge`, which should be identical, but allows called functions to continue to distinguish between `cfg[0]` and `cfg[1]`.

sueloverso commented:
@keithbostic The final changes look good to me.  Given that you added encryption config related changes, should this really be a branch off of our encryption-api branch rather than develop?  Then `secretkey` is already in the `s_string.ok` file and I'm concerned other tests or things on `develop` break when an unknown `encryption=(secretkey=)` appears in the base config file and that doesn't exist on this branch yet?

keithbostic commented:
I think this change should be committed independently of the encryption changes (I could remove the `encryption=(secretkey=)` line, I actually left it in there deliberately to prove I could remove "illegal" strings, since that's one of the reasons I can see us wanting to remove configuration strings in the future.

One of reasons for this change (vs. the original code you and I were talking about), is that empty configuration strings never appear in the base configuration file, so the changes I'm proposing don't alter the base configuration file in any way.

sueloverso commented:
If you want to remove the two references to `secretkey` and then merge to develop, I'm good with that as I'll need to merge it into the branch then and undo the changes I made there.  Thanks!

keithbostic commented:
Why remove the references to `secretkey`, you need them?

I'll do the merge once @michaelcahill approves the approach.

sueloverso commented:
I expect I'll see conflicts since our branch already has them, but I'll see conflicts with conn_api.c no matter what, and it is only one line in `s_string.ok`.  (I thought I was agreeing with your suggestion to remove the reference to secretkey...) :-)

michaelcahill commented:
This seems fine for situations where the the existing `__wt_config_merge` can't easily be used to strip out config keys.  That said, I don't understand why `create` is being explicitly stripped: that should already be handled?

Some encryption config has crept into this branch: if the intent is to merge this into develop, then update encryption, the mentions of "encryption" and "secretkey" should be stripped, and re-added once this change hits the encryption-api branch.

keithbostic commented:
I'm explicitly stripping `create` because the way we did it in the past was to walk `wiredtiger_open_basecfg` (which didn't contain create), and compare it with the list, taking only elements found in `wiredtiger_open_basecfg`. Since we're stripping anyway, it seemed cleaner to use that functionality.

I'll strip out the encryption connections and merge.
